### PR TITLE
The CHANGELOG.

### PR DESCRIPTION
**Pull request by shrit**
We need to be better at keeping the changelog updated :smile:.

Signed-off-by: Omar Shrit <omar@shrit.me>

### CODE REVIEW

File: HISTORY.md [hunk 1]
```diff
@@ -1,5 +1,55 @@
...
+  * Adding a copy constructor in the Convolution layer (#3067).
+
+  * Replace boost::spirit parser by a local efficient implementation
+   (#2942).
+
+  * Disable correctly the autodownloader + fix tests stability (#3076).
+
+  * Replace boost::any with core::v2::any (#3006).
+
+  * Remove old non used Boost headers (#3005)
```
Comment (rcurtin) [inline HISTORY.md:20]:
```suggestion
  * Remove old non used Boost headers (#3005).
```
Sorry for the pedantry. :)

File: HISTORY.md [hunk 2]
```diff
@@ -1,5 +1,55 @@
...
+
+  * Replace boost::enable_if with std::enable_if (#2998).
+
+  * Replace boost::is_same with std::is_same (#2993).
+
+  * Remove non valid option for emsmallen and stb (#2960).
```
Comment (rcurtin) [inline HISTORY.md:26]:
```suggestion
  * Remove invalid option for emsmallen and STB (#2960).
```

File: HISTORY.md [hunk 3]
```diff
@@ -1,5 +1,55 @@
...
+
+  * Check for armadillo dependecies before downloading armadillo (#2954).
```
Comment (rcurtin) [inline HISTORY.md:28]:
```suggestion
  * Check for armadillo dependencies before downloading armadillo (#2954).
```

File: HISTORY.md [hunk 4]
```diff
@@ -1,5 +1,55 @@
...
+
+  * Disable the usage of autodownloader by default (#2953).
+
+  * Install dependecies downloaded with the autodownloader (#2952).
```
Comment (rcurtin) [inline HISTORY.md:32]:
```suggestion
  * Install dependencies downloaded with the autodownloader (#2952).
```

File: HISTORY.md [hunk 5]
```diff
@@ -1,5 +1,55 @@
...
+  * Bump minimum armadillo version from 8.400 to 9.800 (#3043).
+
+  * Adding a copy constructor in the Convolution layer (#3067).
+
+  * Replace boost::spirit parser by a local efficient implementation
+   (#2942).
+
+  * Disable correctly the autodownloader + fix tests stability (#3076).
+
+  * Replace boost::any with core::v2::any (#3006).
```
Comment (rcurtin) [inline HISTORY.md:18]:
```suggestion
  * Replace boost::any with core::v2::any or std::any if available
    (#3006).
```
I think that is correct?  Feel free to ignore the suggestion if it's not.

**Review by rcurtin — COMMENTED** [review]
Awesome, looks good!  I only have a few small comments.

I guess, we should probably have the same goals for our changelog in mlpack that we do for ensmallen, where every PR gets a changelog entry.  A part of keeping the changelog readable there is regular releases---so I guess we will have a *massive* changelog for 4.0.0 when it comes (I will have to make a summary :)), and then we can do regular releases past there. :+1:

File: HISTORY.md (hunk 5 above)
Comment (shrit) [reply HISTORY.md:18]: Yes it is valid for Windows

File: HISTORY.md [hunk 6]
```diff
@@ -1,5 +1,56 @@
 ### mlpack ?.?.?
 ###### ????-??-??
+  * Migrate from boost tests to Catch2 framework (#2523), (#2584).
+
+  * Update NumFOCUS team voting responsibilities (#3071).
```
Comment (zoq) [inline HISTORY.md:5]: I would only list code/documentation changes, I'm not even sure this file is part or should be part of a release.

File: HISTORY.md [hunk 7]
```diff
@@ -1,5 +1,56 @@
...
+
+  * Add armadillo bump version policy (#3048).
+
+  * Bump minimum armadillo version from 8.400 to 9.800 (#3043).
```
Comment (zoq) [inline HISTORY.md:9]: I would move `#3048` into the `Bump minimum armadillo version from 8.400 to 9.800 (#3043).`, similar reason as mentioned above.

File: HISTORY.md [hunk 8]
```diff
@@ -1,5 +1,56 @@
...
+
+  * Adding a copy constructor in the Convolution layer (#3067).
+
+  * Replace boost::spirit parser by a local efficient implementation
+   (#2942).
```
Comment (zoq) [inline HISTORY.md:14]:
```suggestion
  * Replace boost::spirit parser by a local efficient implementation (#2942).
```

I don't think we should use the 80 character rule on the history file.

File: HISTORY.md [hunk 9]
```diff
@@ -1,5 +1,56 @@
...
+
+  * Disable correctly the autodownloader + fix tests stability (#3076).
+
+  * Replace boost::any with core::v2::any or std::any if available
+    (#3006).
```
Comment (zoq) [inline HISTORY.md:19]:
```suggestion
  * Replace boost::any with core::v2::any or std::any if available (#3006).
```

See comment above.

File: HISTORY.md [hunk 10]
```diff
@@ -1,5 +1,56 @@
...
+  * Disable the usage of autodownloader by default (#2953).
+
+  * Install dependencies downloaded with the autodownloader (#2952).
+ 
+  * Download older Boost if the compiler is old (#2940). 
+
+  * Add support for embedded systems (#2531).
+
+  * Build mlpack executable statically if the library is statically linked
+   (#2931).
```
Comment (zoq) [inline HISTORY.md:40]:
```suggestion
  * Build mlpack executable statically if the library is statically linked (#2931).
```

See comment above.

File: HISTORY.md [hunk 11]
```diff
@@ -1,5 +1,56 @@
...
+
+  * Fix a LAPACK bug in FindArmadillo.cmake (#2929).
+
+  * Fix Boost installation on the CI (#2928).
+
+  * Add an autodownloader to get mlpack dependencies (#2927).
+
+  * Remove Coverage files and configurations from CMakeLists (#2866).
+
+  * Remove the double installation of Armadillo in the CI (#2837).
```
Comment (zoq) [inline HISTORY.md:52]: I would remove that one, for the reasons mentioned in the first comment.

File: HISTORY.md [hunk 12]
```diff
@@ -1,5 +1,56 @@
...
+  * Add support for embedded systems (#2531).
+
+  * Build mlpack executable statically if the library is statically linked
+   (#2931).
+
+  * Fix cover tree loop bug on embedded arm systems (#2869).
+
+  * Fix a LAPACK bug in FindArmadillo.cmake (#2929).
+
+  * Fix Boost installation on the CI (#2928).
```
Comment (zoq) [inline HISTORY.md:46]: I would remove that one, for the reasons mentioned in the first comment.

Comment (shrit) [reply HISTORY.md:46]: Okay, so basically remove all lines related to CI, if I understand well.

Comment (zoq) [reply HISTORY.md]: At least that's what I would do, but maybe others like to list everything, @rcurtin what do you think?

Comment (shrit) [reply HISTORY.md]: no worries, I have already removed them,
I think in the future we will have everything listed if things becomes automatic.

File: HISTORY.md [hunk 13]
```diff
@@ -1,5 +1,45 @@
 ### mlpack ?.?.?
 ###### ????-??-??
+  * Migrate from boost tests to Catch2 framework (#2523), (#2584).
+
+  * Bump minimum armadillo version from 8.400 to 9.800 (#3043), (#3048).
+
+  * Adding a copy constructor in the Convolution layer (#3067).
+
+  * Replace boost::spirit parser by a local efficient implementation (#2942).
```
Comment (zoq) [inline HISTORY.md:9]:
```suggestion
  * Replace `boost::spirit` parser by a local efficient implementation (#2942).
```

File: HISTORY.md [hunk 14]
```diff
@@ -1,5 +1,45 @@
...
+
+  * Disable correctly the autodownloader + fix tests stability (#3076).
+
+  * Replace boost::any with core::v2::any or std::any if available (#3006).
```
Comment (zoq) [inline HISTORY.md:13]:
```suggestion
  * Replace `boost::any` with `core::v2::any` or `std::any` if available (#3006).
```

File: HISTORY.md [hunk 15]
```diff
@@ -1,5 +1,45 @@
...
+
+  * Remove old non used Boost headers (#3005).
+
+  * Replace boost::enable_if with std::enable_if (#2998).
```
Comment (zoq) [inline HISTORY.md:17]:
```suggestion
  * Replace `boost::enable_if` with `std::enable_if` (#2998).
```

File: HISTORY.md [hunk 16]
```diff
@@ -1,5 +1,45 @@
...
+
+  * Replace boost::is_same with std::is_same (#2993).
```
Comment (zoq) [inline HISTORY.md:19]:
```suggestion
  * Replace `boost::is_same` with `std::is_same` (#2993).
```

File: HISTORY.md [hunk 17]
```diff
@@ -1,5 +1,45 @@
...
+ 
+  * Download older Boost if the compiler is old (#2940). 
+
+  * Add support for embedded systems (#2531).
+
+  * Build mlpack executable statically if the library is statically linked (#2931).
+
+  * Fix cover tree loop bug on embedded arm systems (#2869).
+
+  * Fix a LAPACK bug in FindArmadillo.cmake (#2929).
```
Comment (zoq) [inline HISTORY.md:37]:
```suggestion
  * Fix a LAPACK bug in `FindArmadillo.cmake` (#2929).
```

**Review by zoq — APPROVED** [review]
Thanks, no further comments from my side.

**Pull request merged**